Site Settings: hook, add and introduce Jetpack Disconnect Site

AnnaMag opened the pull request:
**This is a reference PR.** Cleaned and moved to #18119

Introduce the Disconnect Site, hook it up and add route definitions. Also, the style is added and included for loading. 

This is a landing PR for the new Jetpack Disconnect Flow. It introduces a basic layout of the first page of the flow and makes it visible from the /settings/disconnect-site page (in dev). 

The data layer for the Disconnect Flow is not yet implemented here, nor is saving the survey data.
Style details (e.g. breakpoints) are not fully refined as not all the details are visible yet, so IMHO it makes sense to finalize them in the follow-up PR.

Dev feature flag included.

**To test**:
- Checkout this branch, or use calypso.live
 https://calypso.live/?branch=add/site-settings-disconnect-site

For Jetpack sites (incl. the dev mode):
 - verify that the start of the flow is visible at ```/settings/disconnect-site/:site```, where ```:site``` is the Jetpack site.

For non-Jetpack and Atomic sites, 
 - verify immediate redirection to the General settings

With empty Redux state we should see a placeholder prior to rendering.

**Depends on**: https://github.com/Automattic/wp-calypso/pull/17555

**Visuals**:
![screen shot 2017-09-07 at 10 44 13](https://user-images.githubusercontent.com/13561163/30154235-8d216862-93b9-11e7-9f2d-7e99096f4261.png)

sirreal commented on config/production.json:
```diff
@@ -45,6 +45,7 @@
 		"manage/security": true,
 		"manage/site-settings/analytics": true,
 		"manage/site-settings/categories": true,
+		"manage/site-settings/disconnect-flow": false,
```
MINOR: I believe convention is to omit these until they're activated, but I'm not sure.

If we _do_ add the `false` configs, we should probably add them in the other configs to be consistent.

tyxla replied:
Omitting a flag will simply assume it's disabled in a certain environment, so it's probably best to keep it out in that case.

AnnaMag replied:
removed 👍

sirreal commented on client/my-sites/site-settings/disconnect-site/style.scss:
```diff
@@ -0,0 +1,11 @@
+.disconnect-site__card {
+		width: 70%;
```
This should probably use a [breakpoint], it makes sense on wider displays but probably not on something like a phone. Here's an example:

https://github.com/Automattic/wp-calypso/blob/b599ea715444c909bfd8cd1721c0d160e8b1e18e/client/components/dialog/style.scss#L57-L59

Maybe a designer can provide some better feedback 🙂 

[breakpoint]: https://github.com/Automattic/wp-calypso/blob/master/assets/stylesheets/shared/mixins/_breakpoints.scss

AnnaMag replied:
makes sense 👍

sirreal commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	verifySiteIsJetpack() {
+		if ( this.props.siteIsJetpack === false ) {
```
We're going to need to [check for AT sites](https://github.com/Automattic/wp-calypso/blob/master/client/state/selectors/is-site-automated-transfer.js) as well, they shouldn't be allowed to disconnect.

AnnaMag replied:
moved to a HoC: #17555

sirreal commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,64 @@
+
+	redirectToGeneral = () => {
+		const { siteSlug } = this.props;
+		page( '/settings/general/' + siteSlug );
```
Could we get `/settings/general/undefined` or some other ugly route here?

AnnaMag replied:
This code was moved to an HoC and the Q addressed there: https://github.com/Automattic/wp-calypso/pull/17555

sirreal commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,53 @@
+import { getSelectedSiteSlug } from 'state/ui/selectors';
+
+class PaginationFlow extends Component {
+	verifySiteIsJetpack() {
```
Should this method be here? Looks like it snuck in from the other component.

AnnaMag replied:
great catch

sirreal commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,53 @@
+		}
+	}
+
+	redirectToGeneral = () => {
```
It seems like this is handled in the container component. Is it here just so that the `clickHandler` can do something?

AnnaMag replied:
Yes, this is a redirect for 'Skip survey', which will fast forward the flow. The container component handles permissions. A comment that this is a placeholder would have been a good idea :)

sirreal commented on client/my-sites/site-settings/index.js:
```diff
@@ -1,4 +1,4 @@
-/**
+ /**
```
Bad space snuck in

sirreal commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,64 @@
+		return (
+			<Main className="disconnect-site why site-settings">
+				<DocumentHead title={ translate( 'Site Settings' ) } />
+				<FormattedHeader
```
There are some criteria for which types of sites should see this section, ie is it Jetpack and not Atomic (AT)?

Some places we've started showing placeholders until we have all relevant information to render a component. Do you think it would be a good fit here?

AnnaMag replied:
@tyxla, would you say we still need to use the ```Placeholder``` since we moved the access checks to a HoC ?

tyxla replied:
@AnnaMag yes. Placeholder will be relevant for the case when the user visits the page with a clean Redux state, and will be shown while the sites are loading, similar to how the other settings main sections work.

AnnaMag replied:
Ah yes, now I understand why we need it. It is included now.

sirreal commented:
It would be great to provide some screenshots of the new section in the description so you can see it at a glance.

tyxla commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,64 @@
+		}
+	}
+
+	redirectToGeneral = () => {
```
Let's make sure to not repeat that code multiple times - it's already used in the `ManageConnection` page, and you're introducing new usage of it here. You might be interested in turning it to a HoC that we can take advantege of in several locations.

AnnaMag replied:
addressed in https://github.com/Automattic/wp-calypso/pull/17555

tyxla commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,64 @@
+		return (
+			<Main className="disconnect-site why site-settings">
+				<DocumentHead title={ translate( 'Site Settings' ) } />
+				<FormattedHeader
```
We have a `Placeholder` component in `my-sites/site-settings` that might be handy for that.

tyxla commented on config/stage.json:
```diff
@@ -49,6 +49,7 @@
 		"manage/security": true,
 		"manage/site-settings/analytics": true,
 		"manage/site-settings/categories": true,
+		"manage/site-settings/disconnect-flow": false,
```
Perhaps it would make sense to have it enabled in the staging environment?

nb replied:
Is it working? If it’s not it would be confusing for proxied Automatticians using WordPress.com.

AnnaMag replied:
I guess *not complete* classifies it as *not working*. :)

tyxla commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,53 @@
+	render() {
+		const { translate } = this.props;
+		return (
+			<div className="disconnect-site__pagination main">
```
Did you want to use `components/main` here instead?

tyxla commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,64 @@
+	render() {
+		const { translate } = this.props;
+		return (
+			<Main className="disconnect-site why site-settings">
```
Do we need the `why` class here?

AnnaMag replied:
Not really :)

tyxla commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,64 @@
+					) }
+				/>
+				<div className="disconnect-site__card">
+					<Card> </Card>
```
Do we need a wrapping div, or could we apply the className directly to the `Card`?

tyxla commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,53 @@
+		page( '/settings/general/' + siteSlug );
+	};
+
+	clickHandler = () => {
```
Although `clickHandler` is just fine, and I don't have a strong opinion on how event handlers are named, I'm more inclined to name them like this: `handleClick`.

tyxla commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,53 @@
+				<Button
+					compact
+					borderless
+					className="disconnect-site__pagination skip"
```
Do we need the `.skip` class?

AnnaMag replied:
True, I don't think we need ```className``` here at all (at least for now).

keoshi commented:
Why didn't the bot spin up a calypso.live URL in this PR?

nb commented:
@keoshi because Anna’s not an Automattician – you can use https://calypso.live/?branch=add/site-settings-disconnect-site (may need few minutes to boot)

AnnaMag commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,64 @@
+		}
+	}
+
+	redirectToGeneral = () => {
```
@tyxla, yes, this is a very valid point, which I noted down from your previous comment in the ref PR. I mentioned it in a Note within the PR description. This is The Plan and I believe it is best to be done in general, and not as part of this flow (please correct me if I'm wrong).

AnnaMag replied:
Follow-up thought: this would make a general `redirectTo` component, that e.g. I can use later to redirect to Plans etc

AnnaMag commented:
@keoshi, @rickybanister any thoughts and guidelines on the style issue here (using breakpoints) pointed out by @sirreal ? I searched through the code and WooCommerce login window has a similar appearance http://cld.wthms.co/dqSbhq but its style is not making use of those:
https://github.com/Automattic/wp-calypso/blob/9b6fd9bf368012bf1d11e5f7669bf23e921f8503/client/blocks/login/style.scss
The breakpoint guidelines state to `Think very carefully before adding a new breakpoint`
https://github.com/Automattic/wp-calypso/blob/master/assets/stylesheets/shared/mixins/_breakpoints.scss
so I thought I better check-in with you :)

sirreal commented:
> The breakpoint guidelines state to `Think very carefully before adding a new breakpoint`

The comment is about adding a new and different breakpoint. If you're modifying that code to add a new breakpoint, maybe you shouldn't be 🙂 

As far a _using an existing breakpoint_, you can probably just go for it! See what makes sense. Have a look at this search, you'll see that we _use_ these existing breakpoints in a lot of places, but always from that list of "accepted" breakpoints:

https://github.com/Automattic/wp-calypso/search?l=SCSS&q=breakpoint

AnnaMag commented:
@sirreal, ah true, it is the new ones.
That Woocommerce page resembles a lot the Disconnect landing site, so it made sense to me to look at it as a reference. Yes, I just go for it :)

keoshi commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,44 @@
+		return (
+			<Main className="disconnect-site__pagination main">
+				<Button compact borderless onClick={ this.handleClick }>
+					<Gridicon icon="arrow-right" size={ 18 } />
```
Arrow should be floated to the right of the text.

AnnaMag commented:
All PR reviews are incorporated now.

AnnaMag commented:
I addressed all the points from the reviews, included in a separate commit. That made the size bloated, unfortunately. Hope it is easy to test though 🤞 

@tyxla, @sirreal could you have a look at the final version, please?
@sirreal Re breakpoints: I added some that seemed to make sense, but will further refine it in the subsequent flow, with more designer input. 

Thanks!

tyxla commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,44 @@
+import { getSelectedSiteSlug } from 'state/ui/selectors';
+
+class PaginationFlow extends Component {
+	handleClick = () => {
```
Do we want to use `redirectToGeneral` from the `redirectNonJetpackToGeneral` HoC here? Could be nice, because we wouldn't need to connect this component, and we could convert it to a stateless functional component.

AnnaMag replied:
Ah yes, this is a great idea. It calls for an extension of that HoC's definition to redirect to a prop-passed url, because here we redirect to the next flow steps, not '/general'. Np, since that is the plan anyway 👍

AnnaMag replied:
UPDATE: done https://github.com/Automattic/wp-calypso/pull/17555/

tyxla commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,44 @@
+	render() {
+		const { translate } = this.props;
+		return (
+			<Main className="disconnect-site__pagination main">
```
Do we need the `main` class specifically? I thought the `Main` component adds one by default.

AnnaMag replied:
True! Legacy code from the first iteration.

tyxla commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,52 @@
+	};
+} );
+
+export default flowRight( connectComponent, localize, redirectNonJetpackToGeneral )(
```
Nit: I prefer listing each of those on separate lines, like you've done it for `redirectNonJetpackToGeneral`. This is also valid though, so it's your call.

AnnaMag replied:
@tyxla, I did also here and when I submit the code the lines get merged. The same story happens in other cases at the very end of the files. I suspect that my editor decided to go independent 😂 . 
Seriously speaking though, I will fix it 👍

tyxla commented on client/my-sites/site-settings/disconnect-site/pagination-flow.jsx:
```diff
@@ -0,0 +1,44 @@
+import Main from 'components/main';
+import { getSelectedSiteSlug } from 'state/ui/selectors';
+
+class PaginationFlow extends Component {
```
I'm curious about the component name, `PaginationFlow`, wouldn't it make more sense if we swap the word order,`FlowPagination` or just `Pagination` or even just `SkipSomething` - `SkipSurvey`. I don't see a real pagination except for a skip link, so maybe that would make sense?

AnnaMag replied:
Yes, this definitely could use a better name 👍  `Skip Survey` it is!

tyxla commented on client/my-sites/site-settings/disconnect-site/index.jsx:
```diff
@@ -0,0 +1,52 @@
+	}
+}
+
+const connectComponent = connect( state => {
```
`state => { return { ... } }` can be simplified to `state => ( { ... } )`

AnnaMag commented:
I believe it is better to mark this early code as reference PR as it is too bloated in size and difficult to review.

Split in smaller chunks

johnHackworth commented:
Since this is a reference PR, I'm closing it to avoid any confusion
